File: 39-combination-sum/combination-sum.java
class Solution {
    List<List<Integer>> result = new ArrayList<>();
    List<Integer> ans = new ArrayList<>();

    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        dfs(candidates, 0, target, 0);
        return result;
    }

    void dfs(int[] arr, int i, int target, int sum) {
        if(sum == target) {
            result.add(new ArrayList<>(ans));
            return;
        }
        if(sum > target || i == arr.length){
            return;
        }
        ans.add(arr[i]);
        dfs(arr, i, target, sum + arr[i]);
        ans.remove(ans.size() - 1);
        dfs(arr, i + 1, target, sum);

    }
}

File: 39-combination-sum/combination-sum.py
class Solution:
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        result = []
        ans = []

        def dfs(i, sumT):
            if (sumT == target):
                result.append(ans.copy())
                return
            if (sumT > target or i == len(candidates)):
                return

            ans.append(candidates[i])
            dfs(i, sumT + candidates[i])
            del ans[-1]
            dfs(i + 1, sumT)

        dfs(0, 0)
        return result
